rust/datafusion/src/physical_plan/distinct_expressions.rs

```diff
@@ -0,0 +1,203 @@
+        Ok(Rc::new(RefCell::new(DistinctCountAccumulator {
+            values: FnvHashSet::default(),
+            data_types: self.input_data_types.clone(),
+            count_data_type: self.data_type.clone(),
+        })))
+    }
+}
+
+#[derive(Debug)]
+struct DistinctCountAccumulator {
+    values: FnvHashSet<DistinctScalarValues>,
+    data_types: Vec<DataType>,
+    count_data_type: DataType,
+}
+
+impl Accumulator for DistinctCountAccumulator {
+    fn update_batch(&mut self, arrays: &Vec<ArrayRef>) -> Result<()> {
+        for row in 0..arrays[0].len() {
+            let row_values = arrays
+                .iter()
+                .map(|array| ScalarValue::try_from_array(array, row))
+                .collect::<Result<Vec<_>>>()?;
+
+            self.update(&row_values)?;
+        }
+
+        Ok(())
+    }
+
+    fn update(&mut self, values: &Vec<ScalarValue>) -> Result<()> {
+        // If a row has a NULL, it is not included in the final count.
+        if !values.iter().any(|v| v.is_null()) {
+            self.values.insert(DistinctScalarValues(
+                values
+                    .iter()
+                    .map(GroupByScalar::try_from)
+                    .collect::<Result<Vec<_>>>()?,
+            ));
+        }
+
+        Ok(())
+    }
+
+    fn merge(&mut self, states: &Vec<ScalarValue>) -> Result<()> {
+        self.update(states)
+    }
+
+    fn merge_batch(&mut self, states: &Vec<ArrayRef>) -> Result<()> {
+        let list_arrays = states
+            .iter()
+            .map(|state_array| {
+                state_array.as_any().downcast_ref::<ListArray>().ok_or(
+                    ExecutionError::InternalError(
+                        "Failed to downcast ListArray".to_string(),
+                    ),
+                )
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let values_arrays = list_arrays
+            .iter()
+            .map(|list_array| list_array.values())
+            .collect::<Vec<_>>();
+
+        self.update_batch(&values_arrays)
+    }
+
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        let mut cols_out = self
+            .data_types
+            .iter()
+            .map(|data_type| ScalarValue::List(Some(Vec::new()), data_type.clone()))
+            .collect::<Vec<_>>();
+
+        let mut cols_vec = cols_out
+            .iter_mut()
+            .map(|c| match c {
+                ScalarValue::List(Some(ref mut v), _) => v,
+                _ => unreachable!(),
+            })
+            .collect::<Vec<_>>();
+
+        self.values.iter().for_each(|distinct_values| {
+            distinct_values.0.iter().enumerate().for_each(
+                |(col_index, distinct_value)| {
+                    cols_vec[col_index].push(ScalarValue::from(distinct_value));
+                },
+            )
+        });
+
+        Ok(cols_out)
+    }
+
+    fn evaluate(&self) -> Result<ScalarValue> {
+        match &self.count_data_type {
```

Comment:
This is a good observation. The reason I was explicit with passing the data type around, and validating it here, is because the type computed [here](https://github.com/apache/arrow/blob/6a8f84d5e281e8c20f09642719509fd98c5a81c8/rust/datafusion/src/physical_plan/aggregates.rs#L125) is variable. The reality right now is the type is constant as [UInt64](https://github.com/apache/arrow/blob/6a8f84d5e281e8c20f09642719509fd98c5a81c8/rust/datafusion/src/physical_plan/aggregates.rs#L102), but I wanted to avoid any future issues where the assumption of `UInt64` becomes invalid.